### ISSUE TITLE
Pre-processing Training Data

### ISSUE THREAD

**Issue by lutfia95** [bizerfr/BPNet#3]
Hi, 

I would like to retrain the model with the same training dataset but with other labels. In the paper, you mentioned that you did the pre-processing/labeling using CGAL and OpenCascade data set. I am wondering, how did you generate the training dataset? I am actually interested in 10-18 Labels so I would like to retrain your model again. Thx! 

Best,
Ahmad

**Comment by bizerfr** [comment]
Hi,

The pre-processing is part of an internal toolbox for commercial usage, and we are not allowed to release it at the moment.  But the idea is simple.

1) using OpenCascade to parse the step format of the CAD models. 
2) Operate the Bézier decomposition on the step file. Then, you will get patch-wise rational Bézier patches for each CAD model equipped with uv parameters, control points, patch ids, etc.
3) do random sampling and then fetch the labels.

I attached a TXT(.CPP in fact) file here to show how we extracted the raw labels.  Then, I dump the raw labels into a JSON file. Maybe pythonOCC (python version of OpenCascade) is a better choice.


[test_point_cloud.txt](https://github.com/bizerfr/BPNet/files/13461680/test_point_cloud.txt)

You can implement your own `decomposeToRationalBezierSurfaces`, something like that:

```
void dtkNurbsSurfaceDataOCCT::decomposeToRationalBezierSurfaces(std::vector< std::pair< dtkRationalBezierSurface *, double * > >& r_rational_bezier_surfaces) const
{
    GeomConvert_BSplineSurfaceToBezierSurface convert{d};

    auto nb_u = convert.NbUPatches();
    auto nb_v = convert.NbVPatches();

    TColGeom_Array2OfBezierSurface bezier_surfaces(1, nb_u, 1, nb_v);
    TColStd_Array1OfReal uknots(1, nb_u+1);
    TColStd_Array1OfReal vknots(1, nb_v+2);
    convert.Patches(bezier_surfaces);
    convert.UKnots(uknots);
    convert.VKnots(vknots);

    for (auto i = 1; i <= nb_u; ++i) {
        for (auto j = 1; j <= nb_v; ++j) {
            double* knots = new double[4]; /// @TODO: memory leak
            knots[0] = uknots[i];   // u0
            knots[1] = vknots[j];   // v0
            knots[2] = uknots[i+1]; // u1
            knots[3] = vknots[j+1]; // v1

            if (knots[0] < knots[2] && knots[1] < knots[3]) {
                dtkRationalBezierSurfaceDataOCCT* occt_bezier = new dtkRationalBezierSurfaceDataOCCT();
                auto bezier_surface_handle = bezier_surfaces.Value(i, j);
                occt_bezier->create(bezier_surface_handle);
                r_rational_bezier_surfaces.emplace_back(new dtkRationalBezierSurface{occt_bezier}, knots);
            }
        }
    }
}
```



Best regards,

**Comment by lutfia95** [comment]
Hey, 

sorry for re-asking again, if I understood correctly the labels are not fixed! so the patches id are the labels. By fix labels I mean something like 2 for square ..etc. I am now working on the labeling process, which you mentioned and I recognized this thing. Am I correct? 

Best regards,

**Comment by bizerfr** [comment]
> Hey,
> 
> sorry for re-asking again, if I understood correctly the labels are not fixed! so the patches id are the labels. By fix labels I mean something like 2 for square ..etc. I am now working on the labeling process, which you mentioned and I recognized this thing. Am I correct?
> 
> Best regards,

Hi,

Sorry for the late reply. If you mean the labels for the patch, then it has nothing to do with the primitive types.  If you mean the label for the degeree type, this should be fixed.
